src/settings/toml/migrations.rs

```diff
@@ -0,0 +1,504 @@
+                        old_tag: Some(tag.clone()),
+                        new_tag: Some(tag.clone()),
+                        migration: Migration::default(),
+                    }),
+                    (Some(_), Some(_)) => failure::bail!(
+                        "Multiple migrations with tag {} were found in your wrangler.toml",
+                        tag
+                    ),
+                    (None, _) => failure::bail!(
+                        "The migration with tag {} was not found in your wrangler.toml.",
+                        tag
+                    ),
+                }
+            }
+        }
+    }
+
+    fn coalesce_migrations(
+        migrations: &[MigrationConfig],
+        old_tag: Option<&str>,
+    ) -> Result<ApiMigration, failure::Error> {
+        if migrations.is_empty() {
+            failure::bail!("cannot coalesce empty migrations");
+        }
+        log::info!(
+            "coalescing migrations {:#?} after {:?}",
+            migrations,
+            old_tag
+        );
+        let mut new_classes = migrations[0].migration.durable_objects.new_classes.to_vec();
+        let mut deleted_classes = migrations[0]
+            .migration
+            .durable_objects
+            .deleted_classes
+            .to_vec();
+        let mut renamed_classes = migrations[0]
+            .migration
+            .durable_objects
+            .renamed_classes
+            .to_vec();
+        let mut transferred_classes = migrations[0]
+            .migration
+            .durable_objects
+            .transferred_classes
+            .to_vec();
+
+        for mc in &migrations[1..] {
+            // Renames will modify existing new, rename, and transfer migrations if they are present
+            for rename in mc.migration.durable_objects.renamed_classes.iter() {
+                if let Some(i) = new_classes.iter().position(|c| c == &rename.from) {
+                    new_classes[i] = rename.to.clone();
+                } else if let Some(existing) = renamed_classes
+                    .iter_mut()
+                    .find(|existing| existing.to == rename.from)
+                {
+                    existing.to = rename.to.clone();
+                } else if let Some(existing) = transferred_classes
+                    .iter_mut()
+                    .find(|existing| existing.to == rename.from)
+                {
+                    existing.to = rename.to.clone();
+                } else {
+                    // rename does not modify an existing migration, add it to the final migration
+                    renamed_classes.push(rename.clone());
+                }
+            }
+
+            // deletes will remove existing new, rename, and transfer migrations if they are present
+            for delete in mc.migration.durable_objects.deleted_classes.iter() {
+                if let Some(i) = new_classes.iter().position(|c| c == delete) {
+                    new_classes.remove(i);
```

Comment:
Hm, the more I read and comment on this logic the more worried I am that it's going to be very hard to get right here... What if there already is a class in the script (as previously stored in the API) that has durable objects on it? This will behave very differently than if we actually sent all the migrations to the server -- if we sent one batch at a time, it would be an error to pass a new-class migration for a class that already exists. But if we aggregate them like this, then the upload will silently succeed with the old class still in place.

I'm curious if you have ideas, but I see two options to avoid this sort of problem:

1. When we get the migration tag for the script, also get the list of DO classes currently implemented by the script and then use them to avoid problems like this. We'll have to add this to the API, but that's fine -- it'll be useful for other purposes anyway. What might be a bit less fun is duplicating all of the validation logic here, but it wouldn't be terribly hard.
2. Change the API to support passing an array of migration batches. I'm not sure this would actually reduce the amount of new validation code we need to write, since handling multiple batches at once will require a bunch of new logic will be required in the API -- what's there today isn't really structured for that.

So I think I'd lean towards the first option.

Reply:
Yeah, I'm starting to agree with you as well here. I dislike the idea of re-implementing the validation logic, is there a particular reason you favor 1. over 2.? If it's going to be a similar amount of work, 2. would have the benefit of non-wrangler users getting to upload multiple migrations as well

Reply:
I suppose the main reasons I had in mind are

1. We'll either have to break the existing API (and thus anyone using the current wrangler RC) in order to accept batches of migrations) or put in a hacky compatibility layer in the parsing of the request to deal with some clients sending batches and others assuming that only one batch (not in an array) can be sent, but given that I suppose such a layer is possible it isn't that big of a deal.
2. I don't think it's very likely that anyone other than wrangler will find the option to send multiple rounds of migrations useful, and we'd be making our API a little tougher to use for anyone who only ever sends a single set of migrations at a time.  I don't get the impression we're likely to have many people programming directly against the API, though.

Neither of those points is particularly strong, so I'm left not really caring all that much either way. We can put it in the API if that's your preference.